### PR TITLE
Realtime Bill Yield

### PR DESCRIPTION
**Pull request by elseagle**
This adds the realtime bill yield feature, in place of batch processing which happens in-memory.

### CODE REVIEW

File: openstates/cli/update.py [hunk 1]
```diff
@@ -351,7 +355,11 @@ def _tb_info(type, value, tb):  # type: ignore
 
         sys.excepthook = _tb_info
 
+    logging.info(f"Module: {args.module}")
     juris, module = get_jurisdiction(args.module)
+    logging.info(f"Juris: {juris}")
+    logging.info(f"Module: {module}")
```
Comment (bfossen-ce) [inline openstates/cli/update.py:361]: Could you remove these logging lines if they are no longer needed?

File: openstates/scrape/base.py [hunk 2]
```diff
@@ -138,8 +147,36 @@ def save_object(self, obj):
...
+                with s3.open(S3_FULL_PATH, "w") as file:
+                    # json.dump(obj.as_dict(), file)
+                    json.dump(
+                        OrderedDict(sorted(obj.as_dict().items())),
+                        file,
+                        cls=utils.JSONEncoderPlus,
+                        indent=4,
+                        separators=(",", ": "),
+                    )
+                self.info("S3_FULL_PATH - %s", S3_FULL_PATH)
```
Comment (bfossen-ce) [inline openstates/scrape/base.py:175]: If you could also cleanup the info() lines that would be great

**Review by bfossen-ce — COMMENTED** [review]
Looks good overall. Just need to cleanup some logging and info()

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -45,6 +45,8 @@ jobs:
           poetry-${{ hashFiles('**/poetry.lock') }}
     - name: set poetry config path
       run: poetry config virtualenvs.path ~/.virtualenvs
+    - name: update pip
```
Comment (johnseekins) [inline .github/workflows/test.yml:48]: We don't _need_ this step. Let's take it out.

File: openstates/scrape/base.py [hunk 4]
```diff
@@ -126,6 +134,7 @@ def save_object(self, obj):
         filename = "{0}_{1}.json".format(obj._type, obj._id).replace("/", "-")
 
         self.info("save %s %s as %s", obj._type, obj, filename)
```
Comment (johnseekins) [inline openstates/scrape/base.py:136]: Can we be consistent with string formatting? I don't really care _what_ format, but let's not do c-strings _and_ f-strings in the same file?

File: openstates/settings.py [hunk 5]
```diff
@@ -1,6 +1,9 @@
 import os
 from .utils import transformers
 
+# settings for realtime flag
+S3_REALTIME_BASE = 's3://openstates-realtime-bills/'
```
Comment (johnseekins) [inline openstates/settings.py:5]: This probably should be configurable somehow...having a default is good, having it completely hard-coded is a problem later.

File: pyproject.toml [hunk 6]
```diff
@@ -46,6 +47,7 @@ types-pytz = "^2021.1.0"
 types-click = "^7.1.2"
 types-PyYAML = "^5.4.3"
 types-requests = "^2.25.0"
+s3fs="^2022.10.0"
```
Comment (johnseekins) [inline pyproject.toml:50]: We only need s3fs in the `main` section. Not in dev as well. Duplicate installs can mess things up with poetry.

File: openstates/cli/update.py (hunk 1 above)
Comment (elseagle) [reply openstates/cli/update.py:361]: Done

File: .github/workflows/test.yml (hunk 3 above)
Comment (elseagle) [reply .github/workflows/test.yml:48]: The pipeline was breaking before I added it. I'll remove it now though to ensure it matches devops spec.

File: openstates/scrape/base.py (hunk 4 above)
Comment (elseagle) [reply openstates/scrape/base.py:136]: alright, changed

Comment (johnseekins) [reply openstates/scrape/base.py:135]: Above this comment, we have two distinct string formatting patterns right next to each other. I think that _generally_ OpenStates leans towards f-strings, but as long as the file is at least internally consistent, that works.

File: openstates/scrape/base.py [hunk 7]
```diff
@@ -138,8 +145,33 @@ def save_object(self, obj):
...
 
         if self.scrape_output_handler is None:
-            with open(os.path.join(self.datadir, filename), "w") as f:
-                json.dump(obj.as_dict(), f, cls=utils.JSONEncoderPlus)
+
+            file_path = os.path.join(self.datadir, filename)
+
+            # Remove redundant prefix
+            try:
+                file_path_ = file_path[file_path.index("_data") + 6 :]
```
Comment (johnseekins) [inline openstates/scrape/base.py:153]: Magic number here (`6`). Can you give an example of _why_ you're trimming a portion off the beginning of the string?

File: openstates/scrape/base.py [hunk 8]
```diff
@@ -138,8 +145,33 @@ def save_object(self, obj):
...
+            except Exception:
+                file_path_ = file_path
+
+            if self.realtime:
+
+                s3 = S3FileSystem(anon=False)
+
+                S3_FULL_PATH = settings.S3_REALTIME_BASE + str(file_path_)
```
Comment (johnseekins) [inline openstates/scrape/base.py:161]: This is another string formatting pattern, different than the two above. JS-style str concats are _also_ fine, but we should be consistent.

File: openstates/scrape/base.py [hunk 9]
```diff
@@ -138,8 +145,33 @@ def save_object(self, obj):
...
+
+                with s3.open(S3_FULL_PATH, "w") as file:
```
Comment (johnseekins) [inline openstates/scrape/base.py:163]: When dumping to json for programmatic access, we should avoid adding indents as it is basically just unneeded extra space in the file.

File: openstates/importers/base.py [hunk 10]
```diff
@@ -328,7 +328,8 @@ def import_item(self, data: _JsonDict) -> typing.Tuple[_ID, str]:
         data = self.apply_transformers(data)
         try:
             data = self.prepare_for_db(data)
-        except UnresolvedIdError:
```
Comment (johnseekins) [inline openstates/importers/base.py:331]: Should we catch _all_ errors here? Perhaps the stacktrace was intentional? I know in a lot of cases James designed the systems to fail on exception.

File: openstates/settings.py [hunk 11]
```diff
@@ -1,6 +1,9 @@
 import os
 from .utils import transformers
 
+# settings for realtime flag
+S3_REALTIME_BASE = os.environ.get("S3_REALTIME_BASE")  # e.g 's3://realtime-bucket'
```
Comment (johnseekins) [inline openstates/settings.py:5]: Should we set a default value here? Or will downstream stuff handle an empty string gracefully?

**Review by johnseekins — APPROVED** [review]

**Review by bfossen-ce — APPROVED** [review]

**Pull request merged**